Support vscode as a React Editor

PatoBeltran opened the pull request:
When opening a text editor from react native there are a list of editors being supported. This PR adds `VSCode` to that.

As a difference with the current supported editors, `VSCode` has a notion of `workspace` (a directory that contains all of the project's files). The `workspace` can be passed to `VSCode` as a parameter so that if the workspace is already open we don't get new instances of `VSCode` every time a new file is open. The `workspace` is gotten by comparing the file location with the different `project roots`.

This code relies on `VSCode`'s `code` command, which it's documentation can be found at: https://code.visualstudio.com/Docs/editor/codebasics#_launching-from-the-command-line

facebook-github-bot-7 commented:
By analyzing the blame information on this pull request, we identified @martinbigio and @frantic to be potential reviewers.

facebook-github-bot-1 commented:
@PatoBeltran updated the pull request.

frantic commented:
@PatoBeltran thanks for the PR! I'm a bit concerned about the extra complexity involved regarding the workspaces support – it introduces new couplings that will be hard to maintain in the future.

Instead, you could supply `--reuse-window` argument to `code` which will "Force opening a file or folder in the last active window". That's what Sublime does by default and it's good heuristic in 80% cases.

PatoBeltran commented:
Using `--reuse-window` could be a good approach, the thing is that it wouldn't cover the case when a user has more than one instance of VSCode open (maybe working in different things simultaneously?)

Why are you concerned with the workspace support? The `projectRoot` is already there, we are just passing it through, and the complexity comes from making sure to select the appropriate root from the list provided, which in most cases inside react-native will only be one.

bestander commented on local-cli/server/util/launchEditor.js:
```diff
@@ -51,6 +53,10 @@ function getArgumentsForLineNumber(editor, fileName, lineNumber) {
       return ['--line', lineNumber, fileName];
+    case 'code':
+      var args = ['-g', fileName + ':' + lineNumber];
+      if (workspace) args.unshift(workspace);
```
nit:
please use style
```
if (...) {
...
}
```

bestander commented:
I agree with @frantic that this code introduces a bit of code that is specific to VS code.

If we make custom changes to support VS Code then we need to do the same for Sublime and Idea.
And then this would be hard to support.

@PatoBeltran, do you have a lighter idea, maybe some generic plugin that could trigger any editor however you like?

PatoBeltran commented:
I investigated the idea of using this workspace feature with all of React Native's supported editors. It seems like Nuclide supports this same idea. As far as the other supported editors, Sublime uses workspaces, but when one is specified it opens a new window every time (instead of reusing any open workspace). The other editors (vim, macvim, joe, emacs, textmate, nano) don't support opening a file and a directory at the same time (they only open one of file or folder at a time).

I can implement this feature to work with Nuclide and VSCode, making it a generic feature for any editor that supports workspaces.

bestander commented:
Sounds great!

bestander commented:
Thanks :)

PatoBeltran commented:
@bestander added functionality and addressed comments. This makes any VSCode and Atom based text editors open up with the idea of a workspace. Any future editor that supports this can easily add the workspace to it's arguments to open up properly.

bestander commented:
Great job!
Thanks

bestander commented:
@facebook-github-bot shipit

facebook-github-bot-5 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/238684599832034/int_phab) to review.

bestander commented:
CI infra got broken so I can't see the tests results.
It should be fixed by the time this lands

tlvenn commented:
@PatoBeltran could we add support for vscode insiders (code-insiders command line) as well ?
